src/client.rs

```diff
@@ -241,7 +240,19 @@ impl Read for Encoder {
         }
 
         if !self.body_done {
-            let n = ready!(Pin::new(&mut self.request).poll_read(cx, &mut buf[bytes_read..]))?;
+            let inner_poll_result =
+                Pin::new(&mut self.request).poll_read(cx, &mut buf[bytes_read..]);
+            let n = match inner_poll_result {
+                Poll::Ready(Ok(n)) => n,
+                Poll::Pending => {
+                    if bytes_read == 0 {
+                        return Poll::Pending;
+                    } else {
+                        return Poll::Ready(Ok(bytes_read as usize));
+                    }
+                }
+                e => return e,
```

Comment:
Even though this is more typing, this should be easier to follow since all left-hand match arms then use fully qualified types, rather than a named catch-all argument as is the case now:

```suggestion
                        Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
```